src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Certificate2Collection.cs

```diff
@@ -587,6 +588,164 @@ public bool TryExportCertificatePems(Span<char> destination, out int charsWritte
+
+            OperationStatus status = Convert.FromHexString(thumbprintHex, thumbprint, out _, out int bytesWritten);
+
+            switch (status)
+            {
+                case OperationStatus.InvalidData:
+                case OperationStatus.NeedMoreData:
+                    throw new ArgumentException(SR.Argument_Thumbprint_Invalid, nameof(thumbprintHex));
+                case OperationStatus.DestinationTooSmall:
+                    Debug.Fail("Precomputed buffer was not large enough");
+                    throw new CryptographicException();
+                case OperationStatus.Done:
+                    break;
+            }
+
+            return FindByThumbprintCore(hashAlgorithm, thumbprint.Slice(0, bytesWritten));
+        }
+
+        /// <summary>
+        ///   Searches the collection for certificates with a matching thumbprint.
+        /// </summary>
+        /// <param name="hashAlgorithm">The name of the hash algorithm to compute the thumbprint.</param>
+        /// <param name="thumbprintBytes">The thumbprint to match.</param>
+        /// <returns>A collection of certificates with a matching thumbprint.</returns>
+        /// <exception cref="ArgumentException">
+        ///   <paramref name="hashAlgorithm"/>.<see cref="HashAlgorithmName.Name"/> is <see langword="null"/> or empty.
+        /// </exception>
+        /// <exception cref="PlatformNotSupportedException">
+        ///   <paramref name="hashAlgorithm"/> specifies a hash algorithm not supported by the current platform.
+        /// </exception>
+        /// <exception cref="CryptographicException">
+        /// <para>
+        ///   <paramref name="hashAlgorithm"/> specifies an unknown hash algorithm.
+        /// </para>
+        /// <para>-or-</para>
+        /// <para>
+        ///   An error occured while finding the certificates with a matching thumbprint.
+        /// </para>
+        /// </exception>
+        public X509Certificate2Collection FindByThumbprint(HashAlgorithmName hashAlgorithm, ReadOnlySpan<byte> thumbprintBytes)
+        {
+            ArgumentException.ThrowIfNullOrEmpty(hashAlgorithm.Name, nameof(hashAlgorithm));
+            return FindByThumbprintCore(hashAlgorithm, thumbprintBytes);
+        }
+
+        private X509Certificate2Collection FindByThumbprintCore(HashAlgorithmName hashAlgorithm, ReadOnlySpan<byte> thumbprintBytes)
+        {
+            const int MaxThumbprintStackAlloc = 64; // SHA-2/3-512 is the largest thumbprint currently known.
+            Span<byte> thumbprintBuffer = thumbprintBytes.Length > MaxThumbprintStackAlloc ?
+                new byte[thumbprintBytes.Length] :
+                stackalloc byte[MaxThumbprintStackAlloc];
+
+            X509Certificate2Collection results = [];
+
+            foreach (X509Certificate2 cert in this)
+            {
+                if (!CryptographicOperations.TryHashData(hashAlgorithm, cert.RawDataMemory.Span, thumbprintBuffer, out int bytesWritten))
```

Comment:
_Why doesn't this use `TryGetCertHash`?_

Because that method is virtual. I don't really want to take a contract on calling a virtual. The thumbprint is, by definition, a hash of the certificate.

If we want to take a dependency on it, that's fine, but I am pre-answering why I decided not to use it.